### ISSUE TITLE
C++ exception safety

### ISSUE THREAD

**Issue by vjpai** [grpc/grpc#13706]
Tracking issue to determine whether or not the gRPC core/C++ layer is safe wrt exceptions thrown in application code. Although it is a [Google standard not to throw exceptions](https://google.github.io/styleguide/cppguide.html#Exceptions), we do need to make sure that gRPC doesn't interfere with application writers use of exceptions. We also need to make sure that we don't cause exceptions indirectly

Implications of the above:
1. Don't call libraries that can throw exceptions (e.g., `std::thread`)
    - special case for `new` where an exception is essentially unrecoverable anyway
1. Check places where something must be called before calling a destructor and attempt to remove those
    - e.g., a ServerCompletionQueue must be explicitly drained now before its destructor is called
1. If we call an application-level callback that is not in our control, `catch` any exceptions that it might throw so that we can recover safely
    - e.g., if a service method handlers throws an exception, gRPC C++ should just return with a bad status rather than failing

**Comment by tudor** [comment]
As discussed in #9795, returning an `INTERNAL` status is probably a good default. The programmer can very easily override that if they want a different status code (by writing their own `catch` block), but returning an `INTERNAL` error is arguably better than crashing the server.

(The runtime cost of the `catch` block is zero unless you're actually throwing an exception, unless you're on 32-bit x86 on Windows, which is the only platform that I know of that still uses handler chaining instead of table-based unwinding.)

**Comment by vjpai** [comment]
@tudor: (Just to clarify for those who didn't follow the earlier issue: this is with regard to bullet item 3 listed above) I would prefer UNKNOWN over INTERNAL as the default. According to our status code docs, INTERNAL refers to a breakdown of a system invariant but UNKNOWN means that we don't understand the situation (which is literally true here since the details of the method handler are unknown to the gRPC library). And, like you said, the programmer can write their own blocks if they want.

**Comment by sreecha** [comment]
@vjpai : Can this be closed?

I see that item 2 in your list isn't resolved :
>>2. Check places where something must be called before calling a destructor and attempt to remove those
e.g., a ServerCompletionQueue must be explicitly drained now before its destructor is called

Do you want to keep this issue open till you look through the code for such instances ?

**Comment by vjpai** [comment]
Yes, can you keep it open? I still am not sure that we know the list of such problem places in the API.

**Comment by mmatrosov** [comment]
Hello. Currently, no information about an exception is extracted by the top handler: https://github.com/grpc/grpc/blob/618a3f561d4a93f263cca23abad086ed8f4d5e86/include/grpcpp/impl/codegen/method_handler_impl.h#L43

Wouldn't it be nice to add another catch for `std::exception` and ask its `what()` method to form a more informative error message?

**Comment by vjpai** [comment]
@mmatrosov Hi there, we had originally reported the `what` in Status, but this is a potential security hole if the `what` from some deeply-nested library includes some piece of PII. You're always welcome to define your own exception handlers and use the `what` there, but we shouldn't do it in grpc.

**Comment by uditha-atukorala** [comment]
Hi there, it seems when using the callback API for a server implementation any unhandled exceptions are swallowed and gRPC returns `UNIMPLEMENTED` status code. Potentially this is because of these?:

https://github.com/grpc/grpc/blob/d779808e012d925782b225493c210e268fcd4084/include/grpcpp/support/callback_common.h#L45

Is there a possibility to add an interception hook point to catch any unhandled exceptions?

**Comment by ditsuke** [comment]
I would very much exce

> @mmatrosov Hi there, we had originally reported the `what` in Status, but this is a potential security hole if the `what` from some deeply-nested library includes some piece of PII. You're always welcome to define your own exception handlers and use the `what` there, but we shouldn't do it in grpc as part of the status that gets returned to the client.

That was the right call to make, I would still expect gRPC to at least leave behind a log to prevent total information loss and help in debugging unexpected scenarios.

**Comment by ditsuke** [comment]
> @mmatrosov Hi there, we had originally reported the `what` in Status, but this is a potential security hole if the `what` from some deeply-nested library includes some piece of PII. You're always welcome to define your own exception handlers and use the `what` there, but we shouldn't do it in grpc as part of the status that gets returned to the client.

That was the right call to make, I would still expect gRPC to at least leave behind a log to prevent total information loss and help in debugging unexpected scenarios.
